components/script/dom/window.rs

```diff
@@ -2314,88 +2312,60 @@ impl Window {
         }
     }
 
-    /// Reflows the page if it's possible to do so and the page is dirty. Returns true if layout
-    /// actually happened and produced a new display list, false otherwise.
-    ///
-    /// NOTE: This method should almost never be called directly! Layout and rendering updates
-    /// should happen as part of the HTML event loop via *update the rendering*. Currerntly, the
-    /// only exceptions are script queries and scroll requests.
-    pub(crate) fn reflow(&self, reflow_goal: ReflowGoal, can_gc: CanGc) -> WindowReflowResult {
-        // Never reflow inactive Documents.
-        if !self.Document().is_fully_active() {
-            return WindowReflowResult::new_empty();
+    pub(crate) fn maybe_send_idle_document_state_to_constellation(&self) {
+        if !opts::get().wait_for_stable_image {
+            return;
         }
 
-        // Count the pending web fonts before layout, in case a font loads during the layout.
-        let waiting_for_web_fonts_to_load = self.font_context.web_fonts_still_loading() != 0;
+        if self.has_sent_idle_message.get() {
+            return;
+        }
 
-        self.Document().ensure_safe_to_run_script_or_layout();
+        let document = self.Document();
+        if document.ReadyState() != DocumentReadyState::Complete {
+            return;
+        }
 
-        let updating_the_rendering = reflow_goal == ReflowGoal::UpdateTheRendering;
-        let reflow_result = self.force_reflow(reflow_goal);
+        // Checks if the html element has reftest-wait attribute present.
+        // See http://testthewebforward.org/docs/reftests.html
+        // and https://web-platform-tests.org/writing-tests/crashtest.html
+        if document.GetDocumentElement().is_some_and(|elem| {
+            elem.has_class(&atom!("reftest-wait"), CaseSensitivity::CaseSensitive) ||
+                elem.has_class(&Atom::from("test-wait"), CaseSensitivity::CaseSensitive)
+        }) {
+            return;
+        }
+
+        if self.font_context.web_fonts_still_loading() != 0 {
+            return;
+        }
```

Comment:
thought: Would it make sense to use `fonts.waiting_to_fullfill_promise` here?

Reply:
I agree that this makes sense, though I would prefer to do it in a followup as I'm afraid of introducing too many more behavior changes in this one PR.